### ISSUE TITLE
Docs: document known `no-restricted-syntax` usecases

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#5863]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

In [our FAQ](https://typescript-eslint.io/docs/linting/troubleshooting/#how-can-i-ban-specific-language-feature) we instruct people to use `no-restricted-syntax` for cases of banning specific syntax rather than us creating bespoke rules.

As people report cases, we should add them to the docs so that they're discoverable.
Maybe a sub-page of the FAQ rather than directly on the FAQ itself?

Examples:

Ban `const` enums
```json5
{
  selector: 'TSEnumDeclaration[const=true]',
  message: 'My reason for not using const enums',
}
```

Ban non-`const` enums
```json5
{
  selector: 'TSEnumDeclaration:not([const=true])',
  message: 'My reason for not using const enums',
}
```

Enforce tuple members have labels
```json5
{
  selector: 'TSTupleType > :not(TSNamedTupleMember)',
  message: 'All tuples should have labels',
}
```

Ban `private` members
```json5
{
  selector: ':matches(PropertyDefinition, MethodDefinition)[accessibility="private"]',
  message: 'Use `#private` members instead',
}
```

Ban `#private` members
```json5
{
  selector: ':matches(PropertyDefinition, MethodDefinition) > PrivateIdentifier.key',
  message: 'Use the `private` modifier instead',
}
```

Ban static `this`
```json5
{
  selector: 'MethodDefinition[static = true] ThisExpression',
  message: 'Prefer using the class's name directly',
}
```

### Affected URL(s)

https://typescript-eslint.io/docs/linting/troubleshooting/

**Comment by rajsite** [comment]
Ban accessing `name` on classes and functions:

```js
{
    selector: "MemberExpression[object.property.name='constructor'][property.name='name']"
    message: "'constructor.name' is not reliable (can become 'E', 'P' and etc.) after code minifier usage."
}
```
From [stackoverflow](https://stackoverflow.com/a/72903871) although trying it in a [playground](https://typescript-eslint.io/play/#ts=5.0.2&sourceType=module&code=MYGwhgzhAEBCD2AjaBvaBfAsAKAYgdAHZgC2ApgNw44BmArocAC4CW8h0AsmAE4CeACgCUqDDm78ipSjiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1JFo0fNEgAacNghZJ2SMkRIaw7pACyiALYAjAQFEOABxLJkFfLh74tAK0Q0AdEfwGBpJvdxsNiALwByKnNkclglaD8AXR4nF2g3Dy9fP09vSLEJGShvEzYAc0QVAKCQsITUsApkMAJSMBJ4CjYtJDAACio2XDAdQO8wP10-UX6ABT8wTrowRFIqewBKCYAzWmgwQLpEMA0KXAoligEwWGQ8xHtIDOwAXyuIiVvboA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) doesn't seem to work as intended 😖

**Comment by karlismelderis-mckinsey** [comment]
Ban string type aliases
```ts
{
  selector: "TSTypeAliasDeclaration[typeAnnotation.type='TSStringKeyword']",
  message: "no string type aliases"
}
```

**Comment by JoshuaKGoldberg** [comment]
We'll take this on once #8926 is settled.
